### PR TITLE
[IOTDB-2609] A new lossy encoding method based on frequency domain

### PR DESCRIPTION
**Pull request by 543202718**
This pull request adds a new lossy encoding method named FREQ. The key idea is that the time sequence is transformed to frequency domain and only the frequency components with high energy are reserved. The details of FREQ encoding are shown in https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=199536917

### CODE REVIEW

**Comment by HTHou** [comment]
Hi, please create a Jira issue and link it in the title of this pr. Thanks so much!

File: antlr/src/main/antlr4/org/apache/iotdb/db/qp/sql/IoTDBSqlLexer.g4 [hunk 1]
```diff
@@ -664,6 +664,9 @@ TS_2DIFF
     : T S '_' '2' D I F F
     ;
 
+FREQ
+	: F R E Q
+	;
```
Comment (HTHou) [inline antlr/src/main/antlr4/org/apache/iotdb/db/qp/sql/IoTDBSqlLexer.g4:669]:
```suggestion
FREQ
    : F R E Q
    ;
```

Comment (543202718) [reply antlr/src/main/antlr4/org/apache/iotdb/db/qp/sql/IoTDBSqlLexer.g4:669]: Thanks

File: tsfile/src/main/java/org/apache/iotdb/tsfile/utils/BitConstructor.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+  public BitConstructor() {
+    this.data = new ByteArrayList();
+  }
+
+  public BitConstructor(int initialCapacity) {
+    this.data = new ByteArrayList(initialCapacity);
+  }
+
+  public void add(long x, int len) {
+    x = x & ~(ALL_MASK << len); // 保证x除最低的len位之外都是0
```
Comment (HTHou) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/utils/BitConstructor.java:42]: Use english?

File: tsfile/src/main/java/org/apache/iotdb/tsfile/utils/BitConstructor.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    while (len > 0) {
+      int m = len + cnt >= BITS_IN_A_BYTE ? BITS_IN_A_BYTE - cnt : len; // 向cache中插入的bit数
```
Comment (HTHou) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/utils/BitConstructor.java:44]: english?

File: tsfile/src/main/java/org/apache/iotdb/tsfile/utils/BitConstructor.java [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    return ret;
+  }
+
+  public void clear() {
+    data.clear();
+    cache = 0x00;
+    cnt = 0;
+  }
+
+  /** 如果当前字节存在剩余位，则全部填充0 */
```
Comment (HTHou) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/utils/BitConstructor.java:75]: Same

File: tsfile/src/main/java/org/apache/iotdb/tsfile/utils/BitReader.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    long ret = 0;
+    while (len > 0) {
+      if (bitCnt == BITS_IN_A_BYTE) {
+        next();
+      }
+      int m = len + bitCnt >= BITS_IN_A_BYTE ? BITS_IN_A_BYTE - bitCnt : len; // 从当前byte中读取的bit数
+      len -= m;
+      ret = ret << m;
+      byte y = (byte) (cache & MASKS[bitCnt]); // 运算时byte自动转化为int，需要&截取低位
+      y = (byte) ((y & 0xff) >>> (BITS_IN_A_BYTE - bitCnt - m)); // 逻辑右移，高位补0
```
Comment (HTHou) [inline tsfile/src/main/java/org/apache/iotdb/tsfile/utils/BitReader.java:43]: Use english?

File: server/src/assembly/resources/conf/iotdb-engine.properties [hunk 6]
```diff
@@ -730,6 +730,14 @@ timestamp_precision=ms
...
 # slow_query_threshold=5000
 
+# Signal-noise-ratio (SNR) of FREQ encoding
+# Datatype: double
+# freq_snr=40.0
+
+# Block size of FREQ encoding
+# Datatype: integer
+# freq_block_size=1024
+
```
Comment (HTHou) [inline server/src/assembly/resources/conf/iotdb-engine.properties:740]: Add these parameters to `Reference/Config-Manual.md`.

File: tsfile/src/main/java/org/apache/iotdb/tsfile/utils/BitConstructor.java (hunk 2 above)
Comment (543202718) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/utils/BitConstructor.java]: Rewrite

File: tsfile/src/main/java/org/apache/iotdb/tsfile/utils/BitConstructor.java (hunk 4 above)
Comment (543202718) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/utils/BitConstructor.java]: Rewrite

File: tsfile/src/main/java/org/apache/iotdb/tsfile/utils/BitReader.java (hunk 5 above)
Comment (543202718) [reply tsfile/src/main/java/org/apache/iotdb/tsfile/utils/BitReader.java]: Rewrite

File: server/src/assembly/resources/conf/iotdb-engine.properties (hunk 6 above)
Comment (543202718) [reply server/src/assembly/resources/conf/iotdb-engine.properties:740]: Added

**Comment by HTHou** [comment]
> The details of FREQ encoding are shown in http://ise.thss.tsinghua.edu.cn/confluence/pages/viewpage.action?pageId=64527651

Hi, can you please move this doc to https://cwiki.apache.org/confluence/display/IOTDB/Home? It will be convenient for more people to read.

**Comment by 543202718** [comment]
> > The details of FREQ encoding are shown in http://ise.thss.tsinghua.edu.cn/confluence/pages/viewpage.action?pageId=64527651
> 
> Hi, would you like to move this doc to https://cwiki.apache.org/confluence/display/IOTDB/Home? It will be convenient for more people to read.

That's a good idea. But I don't have the right to edit on this website. Could you please move the doc or give me the right?

**Comment by HTHou** [comment]
> > > The details of FREQ encoding are shown in http://ise.thss.tsinghua.edu.cn/confluence/pages/viewpage.action?pageId=64527651
> > 
> > 
> > Hi, would you like to move this doc to https://cwiki.apache.org/confluence/display/IOTDB/Home? It will be convenient for more people to read.
> 
> That's a good idea. But I don't have the right to edit on this website. Could you please move the doc or give me the right?

Sure. What is your apache confluence ID?

**Comment by 543202718** [comment]
> > > > The details of FREQ encoding are shown in http://ise.thss.tsinghua.edu.cn/confluence/pages/viewpage.action?pageId=64527651
> > > 
> > > 
> > > Hi, would you like to move this doc to https://cwiki.apache.org/confluence/display/IOTDB/Home? It will be convenient for more people to read.
> > 
> > 
> > That's a good idea. But I don't have the right to edit on this website. Could you please move the doc or give me the right?
> 
> Sure. What is your apache confluence ID?

wanghy20

**Comment by HTHou** [comment]
> > Sure. What is your apache confluence ID?
> 
> wanghy20

Done!

**Comment by 543202718** [comment]
The details of FREQ encoding is now also shown in https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=199536917

**Review by HTHou — APPROVED** [review]
LGTM

File: docs/UserGuide/Data-Concept/Encoding.md [hunk 7]
```diff
@@ -54,6 +54,12 @@ Usage restrictions: When using GORILLA to encode INT32 data, you need to ensure
 
 DICTIONARY encoding is lossless. It is suitable for TEXT data with low cardinality (i.e. low number of distinct values). It is not recommended to use it for high-cardinality data. 
 
+* FREQ
+
+FREQ encoding is lossy. It transforms the time sequence to the frequency domain and only reserve part of the frequency components with high energy. It is more suitable for sequence with obvious periodicity.
+
+> There are two parameters of FREQ encoding in the configuration file: `freq_snr` defines the signal-noise-ratio (SNR), `freq_block_size` defines the data size in a time-frequency transformation.
```
Comment (HTHou) [inline docs/UserGuide/Data-Concept/Encoding.md:61]: How do these two parameters affect the encoding performance and compression ratio?

**Review by HTHou — CHANGES_REQUESTED** [review]
Hi, please fix the conflicts and consider add support for Python and C++ client.

See:
https://github.com/apache/iotdb/blob/master/client-py/iotdb/utils/IoTDBConstants.py
https://github.com/apache/iotdb/blob/master/client-cpp/src/main/Session.h

Also, please add tests in 
https://github.com/apache/iotdb/blob/master/integration/src/test/java/org/apache/iotdb/db/integration/IoTDBEncodingIT.java

File: docs/UserGuide/Data-Concept/Encoding.md (hunk 7 above)
Comment (543202718) [reply docs/UserGuide/Data-Concept/Encoding.md:61]: The experimental results and analysis have been put on Apache confluence https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=199536917

Comment (HTHou) [reply docs/UserGuide/Data-Concept/Encoding.md:61]: Yes, I mean, to tell users how to use these parameters, we can show the conclusion in this user guide here. :)

Comment (HTHou) [reply docs/UserGuide/Data-Concept/Encoding.md:61]: Yes, I mean, to tell users how to use these parameters, we can show the conclusion in this user guide here. :)

**Comment by 543202718** [comment]
> Hi, please fix the conflicts and consider add support for Python and C++ client.
> 
> See: https://github.com/apache/iotdb/blob/master/client-py/iotdb/utils/IoTDBConstants.py https://github.com/apache/iotdb/blob/master/client-cpp/src/main/Session.h
> 
> Also, please add tests in https://github.com/apache/iotdb/blob/master/integration/src/test/java/org/apache/iotdb/db/integration/IoTDBEncodingIT.java


Thanks. I have fixed the conflicts and added the support for Python and C++. Meanwhile, the tests are also added in IoTDBEncodingIT.

File: docs/UserGuide/Data-Concept/Encoding.md (hunk 7 above)
Comment (543202718) [reply docs/UserGuide/Data-Concept/Encoding.md]: The brief conclusion is added in the user guide.

**Review by HTHou — APPROVED** [review]
LGTM

**Pull request merged**